intake_xarray/image.py

```diff
@@ -0,0 +1,265 @@
+    ----------
+    files : iter
+        List of file objects
+    imread : function (optional)
+        Optionally provide custom imread function.
+        Function should expect a filename and produce a numpy array.
+        Defaults to ``skimage.io.imread``.
+    preprocess : function (optional)
+        Optionally provide custom function to preprocess the image.
+        Function should expect a numpy array for a single image.
+
+    Examples
+    --------
+    >>> from dask.array.image import imread
+    >>> im = imread('2015-*-*.png')  # doctest: +SKIP
+    >>> im.shape  # doctest: +SKIP
+    (365, 1000, 1000, 3)
+
+    Returns
+    -------
+    Dask array of all images stacked along the first dimension.  All images
+    will be treated as individual chunks
+    """
+    from dask.array import Array
+    from dask.base import tokenize
+    from dask.bytes.core import OpenFile
+
+    if not imread:
+        from skimage.io import imread
+
+    def _imread(open_file: OpenFile):
+        with open_file as f:
+            return imread(f)
+
+    filenames = [f.path for f in files]
+
+    name = 'imread-%s' % tokenize(filenames, map(get_mtime, filenames))
+
+    with files[0] as f:
+        sample = imread(f)
+    if preprocess:
+        sample = preprocess(sample)
+
+    keys = [(name, i) + (0,) * len(sample.shape)
+            for i in range(len(files))]
+
+    if preprocess:
+        values = [(add_leading_dimension, (preprocess, (_imread, f)))
+                  for f in files]
+    else:
+        values = [(add_leading_dimension, (_imread, f))
+                  for f in files]
+    dsk = dict(zip(keys, values))
+
+    chunks = ((1, ) * len(files), ) + tuple((d, ) for d in sample.shape)
+
+    return Array(dsk, name, chunks, sample.dtype)
+
+
+def reader(file, chunks, imread=None, preprocess=None, **kwargs):
+    """ Read a file object and output an xarray object
+    """
+    import numpy as np
+    from xarray import DataArray
+
+    if not imread:
+        from skimage.io import imread
+
+    with file as f:
+        array = imread(f)
+    if preprocess:
+        array = preprocess(array)
+
+    ny, nx = array.shape[:2]
+    coords = {'y': np.arange(ny),
+              'x': np.arange(nx)}
+    dims = ('y', 'x')
+
+    if len(array.shape) == 3:
+        nband = array.shape[2]
+        coords['band'] = np.arange(nband)
+        dims += ('band',)
+
+    return DataArray(array, coords=coords, dims=dims).chunk(chunks=chunks)
+
+
+def multireader(files, chunks, concat_dim, **kwargs):
+    import numpy as np
+    from xarray import DataArray
+
+    dask_array = dask_imread(files, **kwargs)
+
+    ny, nx = dask_array.shape[1:3]
+    coords = {'y': np.arange(ny),
+              'x': np.arange(nx)}
+    if isinstance(concat_dim, list):
+        dims = ('dim_0', 'y', 'x')
+    else:
+        dims = (concat_dim, 'y', 'x')
+
+    if len(dask_array.shape) == 4:
+        nband = dask_array.shape[3]
+        coords['band'] = np.arange(nband)
```

Comment:
"band" sounds rather specific to the application here. Is this RGB/channels or something else?

Reply:
In datashader we used the term "layer" to avoid the specific connection to LANDSAT bands; many types of image files have layers.

Reply:
It must b e the Xray astronomer in me that wants to say "channel" :)

Reply:
We also considered "channel" for datashader, given that R, G, B, and A are often called channels, but layer seemed more general because it also covers image stacks all kept in the same file distinguished by their z-value.

Reply:
ok so maybe channel is right here. This is for regular images.